### ISSUE TITLE
yolo layer configuration

### ISSUE THREAD

**Issue by pfeatherstone** [AlexeyAB/darknet#5168]
I've noticed their are some additional configurations to the yolo layer, which are used in the csresnext model:

```
truth_thresh = 1
scale_x_y = 1.1
iou_thresh=0.213
cls_normalizer=1.0
iou_normalizer=0.07
```

what do these do? Particularly scale_x_y

**Comment by pfeatherstone** [comment]
am i correct in saying rather than having 
```bx[i] = sigmoid(tx[i]) + c[i]```
you get
```bx[i] = (sigmoid(tx[i]) - 0.5) * scale_x_y + 0.5 + c[i]```

**Comment by AlexeyAB** [comment]
@pfeatherstone Yes. There are many experimental params.

**Comment by pfeatherstone** [comment]
Do class probabilities or width and height get scaled?

**Comment by pfeatherstone** [comment]
Basically I'm doing a port of csresnext50-panet-spp-original-optimal.cfg to pytorch. I've got all the layers down but I think there are a few fiddly things which are added in csresnext50-panet-spp-original-optimal.cfg which you wouldn't have in yolov3.cfg. When inferring I get correct-ish bounding boxes but they are not exactly the same as a darknet inferrence. So i think there are some additional subtleties missing.

**Comment by AlexeyAB** [comment]
* Do you get the same bbox coords for yolov3-spp.cfg but different coords for csresnext50-panet.cfg?
* Did you successfully implemented: grouped-convolutional, scale_x_y, partial shortcut?

About shortcut:

![72104147-d225c880-333b-11ea-8692-b0b1b1b86bb8](https://user-images.githubusercontent.com/4096485/78379943-d16b9e00-75db-11ea-8af8-3b3df15c5999.png)

![image](https://user-images.githubusercontent.com/4096485/78379887-bd27a100-75db-11ea-8b51-480e8ecab24c.png)

**Comment by pfeatherstone** [comment]
i get slightly different bboxes for **both** yolov3-spp and csresnext50... when comparing against your darknet repo and my pytorch implementation. yolov3 on the other hand i get **exactly** the same results. yolov3-spp doesn't have any partial shortcuts, so i'm inclined to think it's not that. I was hoping it was a new experimental scaling somewhere

**Comment by pfeatherstone** [comment]
for yolov3-spp i get the following 
![predictions](https://user-images.githubusercontent.com/45853521/78380911-84cc9680-75cc-11ea-8992-ae5bdbe86aba.jpg)
![predictions2](https://user-images.githubusercontent.com/45853521/78380923-87c78700-75cc-11ea-8642-f8d305962c1c.jpg)

The first image is what i get with your repo, the second is what i get with my pytorch implementation

**Comment by pfeatherstone** [comment]
Both are inferred on 416x416

**Comment by pfeatherstone** [comment]
for csresnext50-panet-spp-original-optimal.cfg i get
![predictions](https://user-images.githubusercontent.com/45853521/78381276-0a504680-75cd-11ea-955c-3c1c604658b9.jpg)
![predictions2](https://user-images.githubusercontent.com/45853521/78381282-0d4b3700-75cd-11ea-872e-993358c133f7.jpg)

both are inferred on 416 x 416

**Comment by pfeatherstone** [comment]
It's so similar, it makes me think it's a scaling thing, or a different mode in the upsampling layers or something like that.

**Comment by pfeatherstone** [comment]
Maybe it's the maxpooling. that's the only difference between yolov3 and yolov3-spp.

**Comment by pfeatherstone** [comment]
maybe darknet has a slightly different implementation to pytorch

**Comment by AlexeyAB** [comment]
* Do you get the same bbox for yolov3.cfg but different for yolov3-spp.cfg?

* Check SPP-block. And do you use `padding=SAME` in Pytorch for maxpool layers in SPP-block? https://github.com/pytorch/pytorch/issues/3867#issuecomment-570743193

* Use the same resized to 416x416 `dog.png` image, not jpeg, and not 768x576. Resize it in Paint and save to dog.png. May be you are using different resizing approach: https://github.com/AlexeyAB/darknet/issues/232#issuecomment-336955485

* Also check that you use get the same output before NMS in both cases Darknet and Pytorch.

**Comment by pfeatherstone** [comment]
Yes I get exactly the same bbox for yolov3.cfg but different for yolov3-spp.cfg

**Comment by pfeatherstone** [comment]
The maxpool layers in pytorch use zero-padding

**Comment by pfeatherstone** [comment]
Ah yes, i'm not doing letterbox resizing in pytorch

**Comment by AlexeyAB** [comment]
Use `-letter_box` flag in Darknet to detect with letter_box resizig.

`./darknet detector test ... -letter_box`

----

What is the padding=SAME and padding=VALID: https://www.pico.net/kb/what-is-the-difference-between-same-and-valid-padding-in-tf-nn-max-pool-of-tensorflow


----

Padding SAME and VALID in TensorFlow: https://www.tensorflow.org/api_docs/python/tf/nn/convolution

> If padding == "SAME": output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides[i])
> 
> If padding == "VALID": output_spatial_shape[i] = ceil((input_spatial_shape[i] - (spatial_filter_shape[i]-1) * dilation_rate[i]) / strides[i]).

----

https://github.com/pytorch/pytorch/issues/3867#issuecomment-570743193

> Still no padding='same' in Pytorch

**Comment by pfeatherstone** [comment]
So here is the the input image
![dog](https://user-images.githubusercontent.com/45853521/78384027-569d8580-75d1-11ea-8949-94d8365123ff.png)
here is your prediction
![predictions](https://user-images.githubusercontent.com/45853521/78384053-60bf8400-75d1-11ea-9d30-635be1b12108.jpg)
here is my prediction
![prediction2](https://user-images.githubusercontent.com/45853521/78384079-6d43dc80-75d1-11ea-9201-ac7e0424c904.png)
You get the following class probabilities:
bicycle: 86%
bicycle: 53%
dog: 98%
car: 26%
truck: 90%
Mine are painted on the image

**Comment by AlexeyAB** [comment]
Remove SPP-block from yolov3-spp.cfg in both cases and run with the same yolov3-spp.weights, do you get the same results?

**Comment by pfeatherstone** [comment]
There is an extra convolutional layer following the SPP block. It takes 2048 channels

**Comment by pfeatherstone** [comment]
When inferring with your repo i get no boxes. I image it's because the weights are offset from removing the SPP block and the following convolutional layer taking a different number of channels

**Comment by AlexeyAB** [comment]
> When inferring with your repo i get no boxes. I image it's because the weights are offset from removing the SPP block and the following convolutional layer taking a different number of channels

Yes, it will not work.

I think the issue - you don't use padding=SAME: https://github.com/AlexeyAB/darknet/issues/5168#issuecomment-608539941

**Comment by pfeatherstone** [comment]
Yeah padding=SAME in my stuff. Otherwise i wouldn't end up with 52x52, 26x26 and 13x13 as the input dimensions of the yolo layers

**Comment by pfeatherstone** [comment]
And i've checked, it pads in 'replicate' mode, not 'zeros'

**Comment by AlexeyAB** [comment]
1. Check detections without NMS

2. It should be `padding=SAME`
![Figure2 1581475727](https://user-images.githubusercontent.com/4096485/78389838-20b9ca80-75ec-11ea-8dd1-95b969477c06.gif)

**Comment by pfeatherstone** [comment]
Cheers @AlexeyAB. I'm convinced the pooling is correct. Maybe it's something else. I'll have a look tomorrow. Thanks for your time!
